bentoml/saved_bundle/templates.py

```diff
@@ -68,25 +68,33 @@
 graft {service_name}/artifacts
 """
 
+BUNDLED_PIP_DEPENDENCIES_DOCKER_COPY = """\
+COPY bundled_pip_dependencies /bento/bundled_pip_dependencies
+"""
+
 MODEL_SERVER_DOCKERFILE_CPU = """\
 FROM {docker_base_image}
 
-# copy over model files
-COPY . /bento
-WORKDIR /bento
-
 # Configuring PyPI index
 ARG PIP_INDEX_URL=https://pypi.python.org/simple/
 ARG PIP_TRUSTED_HOST=pypi.python.org
 ENV PIP_INDEX_URL $PIP_INDEX_URL
 ENV PIP_TRUSTED_HOST $PIP_TRUSTED_HOST
 
+# copy over files needed for init script
+COPY environment.yml requirements.txt setup.sh* bentoml-init.sh /bento/
```

Comment:
what's the reason for the `*` in `setup.sh*`?

Reply:
It seems to me we could just do the following here:
```
COPY environment.yml requirements.txt setup.sh bentoml-init.sh bundled_pip_dependencies* /bento/
```

And avoid the changes around `bundled_pip_dependencies_saved` and make the code path a little cleaner. Is there any reason that this may not work or it will invalidate the build cache?  For users, the `bundled_pip_dependencies` directory should be empty. This directory is only for BentoML developers to bundle their local fork of BentoML in the docker image.

Reply:
@parano Thank you for such a quick response.

I tried a lot to make it something similar but I faced some limitations in docker unfortunately.

Since the ```setup.sh``` file is optional, it is not always included in the service directory, so docker gives an error when it tries to copy it and it does not exist.
The ```*``` is a workaround to allow docker to skip it if it does not exist.

For the ```bundled_pip_dependencies```, I agree the changes are not clean at all.
The problem I faced when it is copied this way is that it is flattened in the ```/bento/``` directory not in its own directory (i.e. ```/bento/bundle.tar.gz``` not ```/bento/bundled_pip_dependencies/bundle.tar.gz```), which is not compatible with ```bento-init.sh``` script.

I am thinking of two ways to make it cleaner and avoid changes around ```bundled_pip_dependencies_saved``` altogether:
1. Change ```bento-init.sh``` to loop over the flattened files ```for filename in ./*.tar.gz``` but I am afraid this could break the script if someone is using it locally not in the context of docker but it could just be changed try both directories ```./*.tar.gz``` and ```./bundled_pip_dependencies/*.tar.gz``` and that would make it work in both contexts.
2. After copying the ```*.tar.gz``` into  ```/bento/``` copy them into the correct directory ```RUN mkdir bundled_pip_dependencies && mv /bento/*.tar.gz /bento/bundled_pip_dependencies/```

What do you think? your feedback would be very helpful.

Reply:
Got it, that makes sense about `setup.sh`

As for `bundled_pip_dependencies`, maybe try add the following line after the `COPY . /bento` line, and remove related code from `bentoml-init.sh`?

```
RUN if [ -d /bento/bundled_pip_dependencies ]; then pip install -U bundled_pip_dependencies/* ;fi
```

Reply:
Nice, This is very elegant.